test/unit/BPool/BPool_Bind_Unbind.t.sol

```diff
@@ -0,0 +1,207 @@
+    // it should revert
+    vm.expectRevert(IBPool.BPool_TokensAboveMaximum.selector);
+    bPool.bind(token, tokenBindBalance, tokenWeight);
+  }
+
+  function test_BindRevertWhen_TokenWeightIsTooLow() external whenReentrancyLockIsNOTSet whenCalledByController {
+    // it should revert
+    vm.expectRevert(IBPool.BPool_WeightBelowMinimum.selector);
+    bPool.bind(token, tokenBindBalance, MIN_WEIGHT - 1);
+  }
+
+  function test_BindRevertWhen_TokenWeightIsTooHigh() external whenReentrancyLockIsNOTSet whenCalledByController {
+    // it should revert
+    vm.expectRevert(IBPool.BPool_WeightAboveMaximum.selector);
+    bPool.bind(token, tokenBindBalance, MAX_WEIGHT + 1);
+  }
+
+  function test_BindRevertWhen_TooLittleBalanceIsProvided() external whenReentrancyLockIsNOTSet whenCalledByController {
+    // it should revert
+    vm.expectRevert(IBPool.BPool_BalanceBelowMinimum.selector);
+    bPool.bind(token, MIN_BALANCE - 1, tokenWeight);
+  }
+
+  function test_BindRevertWhen_WeightSumExceedsMAX_TOTAL_WEIGHT()
+    external
+    whenReentrancyLockIsNOTSet
+    whenCalledByController
+  {
+    bPool.set__totalWeight(2 * MAX_TOTAL_WEIGHT / 3);
+    // it should revert
+    vm.expectRevert(IBPool.BPool_TotalWeightAboveMaximum.selector);
+    bPool.bind(token, tokenBindBalance, MAX_TOTAL_WEIGHT / 2);
+  }
+
+  modifier whenPoolIsNOTFinalized() {
+    _;
+  }
+
+  function test_BindWhenItIsNOTFinalized() external whenReentrancyLockIsNOTSet whenCalledByController {
+    // it calls _pullUnderlying
+    bPool.expectCall__pullUnderlying(token, deployer, tokenBindBalance);
+    // it sets reentrancy lock
+    bPool.expectCall__setLock(_MUTEX_TAKEN);
+
+    bPool.bind(token, tokenBindBalance, tokenWeight);
+
+    // it adds token to the list
+    assertEq(bPool.call__tokens()[0], token);
+    // it sets the token record
+    assertEq(bPool.call__records(token).bound, true);
+    assertEq(bPool.call__records(token).denorm, tokenWeight);
+    // it sets total weight
+    assertEq(bPool.call__totalWeight(), tokenWeight);
+  }
+
+  function test_BindRevertWhen_ReentrancyLockIsSet() external {
+    bPool.call__setLock(_MUTEX_TAKEN);
+    vm.expectRevert(IBPool.BPool_Reentrancy.selector);
+    // it should revert
+    bPool.bind(token, tokenBindBalance, tokenWeight);
+  }
+
+  function test_UnbindRevertWhen_CalledNOTByController(address _caller) external whenReentrancyLockIsNOTSet {
+    // it should revert
+    vm.assume(_caller != deployer);
+    vm.prank(_caller);
+    vm.expectRevert(IBPool.BPool_CallerIsNotController.selector);
+    bPool.unbind(token);
+  }
+
+  function test_UnbindRevertWhen_TokenIsNotBound() external whenReentrancyLockIsNOTSet whenCalledByController {
+    vm.expectRevert(IBPool.BPool_TokenNotBound.selector);
+    // it should revert
+    bPool.unbind(token);
+  }
+
+  modifier whenTokenIsBound() {
+    // note: it depends on the `startPrank` of whenCalledByController
+    bPool.bind(token, tokenBindBalance, tokenWeight);
```

Comment:
why not using `set__record`?

Reply:
because I would also need to set the total weight? but yeah I can also set that, thanks for pointing it out